### PR TITLE
11679 BigQuery-Denormalized Destination: improve code coverage

### LINKED ISSUE #11679: BigQuery-Denormalized Destination: improve code coverage
### Acceptance criteria
- [ ] unit test coverage should be about 90%

### PR DESCRIPTION
**Pull request by suhomud**
## What
Improve code coverage for BigQuery-Denormalized Destination unit test
**Edited**
The initial task was to improve the code coverage for unit tests however with existing structure of the unit test for BigQuery-Denormalized Destination it was hard to achieve. In most cases our unit tests for connectors are really similar to integration test. 
NOTE: 
- There is no logic change for the connector but only refactoring for code duplication and clean up for unused code
- The integration tests are not changed 

## How 
The idea of this refactoring is to have a separate unit test class for each implementation class which needs to be tested. If unit tests duplicates implementation structure we have some benefits like access to protected methods and support, extension and readability of the unit tests. 

### Change list explanation
 
- Unused method [here](https://github.com/airbytehq/airbyte/pull/17827/files#diff-9a951f660091f354c15d85e748559735385e2d696d303927b4ecea0273641889L62) and [here](https://github.com/airbytehq/airbyte/pull/17827/files#diff-981ded32f449885879d1af3dd682aa63b05f30687b6d35c6597979b91580811dL84) were removed
- Code duplication [here](https://github.com/airbytehq/airbyte/pull/17827/files#diff-de99e147aaa0a078a591a1941617f384e54b7cba76062baf2474d808e04410d4L68) and [here](https://github.com/airbytehq/airbyte/pull/17827/files#diff-1e413d263a9871268a23cedeb7db13cd98899ee7ce93c25592fa28edf6db3437L42). This functionality is moved to [FormatterUtil](https://github.com/airbytehq/airbyte/pull/17827/files#diff-da0f3e2f4f705bc5caed2075ca5613281788310d799f6a50862fa9d621ae4837R16) class
- [TestBigQueryDenormalizedRecordFormatter](https://github.com/airbytehq/airbyte/pull/17827/files#diff-496ecb3683c64e24db6ab7e365db0a1d1d9b9d8db2dc832c4c3bbca8dbfa2e35) and [TestGcsBigQueryDenormalizedRecordFormatter](https://github.com/airbytehq/airbyte/pull/17827/files#diff-38c29ecc822fdce73f963beda98d8c909583d6f8bcc17bde01d06cffc7f6e554) classes used only to make `protected void addAirbyteColumns(final ObjectNode data, final AirbyteRecordMessage recordMessage)` method as `public`. After refactoring protected method can be accessed from unit test
- The previous test BigQueryDenormalizedUtilsTest was not changed but splited to DefaultBigQueryDenormalizedRecordFormatterTest and GcsBigQueryDenormalizedRecordFormatterTest
- new unit tests structure:
Screenshot 2022-10-13 at 10 43 30

### CODE REVIEW

**Review by etsybaev — APPROVED** [review]

**Comment by edgao** [comment]
is there a recommended reading order for the files in this PR? Also, It looks like some of this is refactoring / moving code around, could you summarize those changes?

(would be great to put this info in the PR description)

**Comment by suhomud** [comment]
> is there a recommended reading order for the files in this PR? Also, It looks like some of this is refactoring / moving code around, could you summarize those changes?
> 
> (would be great to put this info in the PR description)

@edgao sorry it's definitely should be explained. I added PR description please have a look

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    when(tableMock.getDefinition()).thenReturn(tableDefinitionMock);
+
+    // expected schema is null
+    when(bigQueryRecordFormatterMock.getBigQuerySchema()).thenReturn(null);
+    // existing schema is null
+    when(tableDefinitionMock.getSchema()).thenReturn(null);
+    // run test
+    bqdd.putStreamIntoUploaderMap(airbyteStreamMock, uploaderConfigMock, uploaderMap);
+    // should not use LegacyArrayFormatter
+    verify(bigQueryRecordFormatterMock, times(0)).setArrayFormatter(any(LegacyArrayFormatter.class));
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:176]: what _should_ this use?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+  @Test
+  void putStreamIntoUploaderMap_compareSchemas_expectedIsNotNullExistingIsNull() throws IOException {
+    final Map<AirbyteStreamNameNamespacePair, AbstractBigQueryUploader<?>> uploaderMap = new HashMap<>();
+    final String streamName = "stream_name";
+    final String nameSpace = "name_space";
+    final Table tableMock = mock(Table.class);
+    final Schema schemaMock = mock(Schema.class);
+    final TableDefinition tableDefinitionMock = mock(TableDefinition.class);
+
+    when(uploaderConfigMock.getConfigStream()).thenReturn(configuredStreamMock);
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:133]: it looks like this setup code is repeated a lot, can we extract it into a method?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  @Test
+  void fromJsonSchemaFormat_matchByFormat() {
+    JsonSchemaFormat result = JsonSchemaFormat.fromJsonSchemaFormat("date", null);
+    assertEquals(JsonSchemaFormat.DATE, result);
+  }
+
+  @Test
+  void fromJsonSchemaFormat_notExistingFormat() {
+    JsonSchemaFormat result = JsonSchemaFormat.fromJsonSchemaFormat("not_existing_format", null);
+    assertNull(result);
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java:28]: probably unrelated to your change, but I feel like this should default to string rather than null - that way we always have safe behavior, even for weird source schemas

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+
+  @Test
+  void testSchema() {
+    final JsonNode jsonNodeSchema = getSchema();
+    DefaultBigQueryDenormalizedRecordFormatter rf = new DefaultBigQueryDenormalizedRecordFormatter(
+        jsonNodeSchema, new BigQuerySQLNameTransformer());
+
+    final Schema bigQuerySchema = rf.getBigQuerySchema(jsonNodeSchema);
+
+    final FieldList fields = bigQuerySchema.getFields();
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java:80]: is it possible to construct an `expectedSchema` object and just `assertEquals(expectedSchema, bigQuerySchema)`? This approach looks pretty verbose and easy to introduce errors (or accidentally forget to assert something)

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java [hunk 5]
```diff
@@ -248,8 +260,8 @@ public void testEmittedAtTimeConversion() {
     airbyteRecordMessage.setEmittedAt(1602637589000L);
     mockedFormatter.addAirbyteColumns(objectNode, airbyteRecordMessage);
 
-    assertEquals("2020-10-14 01:06:29.000000+00:00",
-        objectNode.get(JavaBaseConstants.COLUMN_NAME_EMITTED_AT).textValue());
+    assertEquals("1602637589000",
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java:263]: why did this value change? Shouldn't we still be emitting ISO8601 timestamps?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/DefaultArrayFormatterTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+  @Test
+  void formatArrayItems() {
+    JsonNode arrayFirst = mapper.createArrayNode().add("one").add("two");
+    JsonNode arraySecond = mapper.createArrayNode().add("one").add("two");
+    List<JsonNode> arrayItems = List.of(arrayFirst, arraySecond);
+
+    JsonNode result = formatter.formatArrayItems(arrayItems);
+
+    assertNotNull(result);
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/DefaultArrayFormatterTest.java:40]: how about something like this:
```java
JsonNode expectedArrayNode = mapper.readTree(
  """
  [
    {"big_query_array": ["one", "two"]},
    {"big_query_array": ["one", "two"]},
  ]
  """
);
assertEquals(expectedArrayNode, result);
```

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/DefaultArrayFormatterTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    assertEquals(objectNode, result.get(0));
+  }
+
+  @Test
+  void findArrays() {
+    JsonNode schemaArrays = getSchemaArrays();
+
+    List<JsonNode> result = formatter.findArrays(schemaArrays);
+
+    assertNotNull(result);
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/DefaultArrayFormatterTest.java:65]: similar question as above, can this just assert against an expected list of json nodes?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/util/FormatterUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.Test;
+
+class FormatterUtilTest {
+
+  private final ObjectMapper mapper = new ObjectMapper();
+
+  @Test
+  void isAirbyteArray_typeIsNull() {
+    JsonNode arrayNode = mapper.createArrayNode().add("one").add("two");
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/util/FormatterUtilTest.java:21]: I think these tests would all be easier to read if they used a literal json string:
```java
arrayNode = mapper.readTree("""["one", "two"]""");

arrayNode = mapper.readTree(
  """
  {
    "type": ["array"]
  }
  """);

// etc
```

**Review by edgao — COMMENTED** [review]
thanks for the description, was super helpful for reading through this!

I added some comments - overall the structure makes a lot of sense, so I just focused on general test implementations. Didn't read through the actual test cases in a lot of depth (LMK if you'd find that valuable); I was more looking for ways to improve readability + future  maintainability.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+import io.airbyte.integrations.destination.bigquery.BigQuerySQLNameTransformer;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+
+class DefaultBigQueryDenormalizedRecordFormatterTest {
```
[the remark this answers is not in the document]
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java:48]: It test DefaultBigQueryDenormalizedRecordFormatter so the Standard name can be confused. The idea is - unit test should be in align with class name which it test

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+  public void teardown() {
+    uploaderFactoryMock.close();
+  }
+
+  @Test
+  void getTargetTableName() {
+    final String streamName = "stream_name";
+    BigQuerySQLNameTransformer nameTransformerMock = mock(BigQuerySQLNameTransformer.class);
+    bqdd.getTargetTableName(streamName);
+    verify(nameTransformerMock, times(0)).getIdentifier(streamName);
```
[the remark this answers is not in the document]
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:91]: Good question. Probably I was thinking how should I test it but just forgot to edit it 😕. The method `getTargetTableName` in BigQueryDenormalizedDestination class does nothing but only calling namingResolver.getIdentifier(streamName) which is created in default constructor. 

I updated it to test the expected result but question for me is should we test functionality from others classes like `BigQuerySQLNameTransformer` to be able to fill requirement of 90 % coverage since in this case we test this functionality multiple times in a different Unit tests

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+  }
+
+  @Test
+  void isDefaultAirbyteTmpTableSchema() {
+    assertFalse(bqdd.isDefaultAirbyteTmpTableSchema());
+  }
+
+  @Test
+  void getAvroSchemaCreator() {
+    assertNotNull(bqdd.getAvroSchemaCreator());
```
[the remark this answers is not in the document]
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:112]: Same here. I think we should not test the functionality of JsonToAvroSchemaConverter().getAvroSchema here. So I I'll verify that lambda used arguments of mocks here

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+  }
+
+  @Test
+  void getRecordFormatterCreator() {
+    final BigQuerySQLNameTransformer nameTransformerMock = mock(BigQuerySQLNameTransformer.class);
+    final BigQueryRecordFormatter resultFormatter = bqdd.getRecordFormatterCreator(nameTransformerMock)
+        .apply(mock(JsonNode.class));
+
+    assertThat(resultFormatter, instanceOf(GcsBigQueryDenormalizedRecordFormatter.class));
```
[the remark this answers is not in the document]
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:121]: Mostly I did it to cover almost all methods but the `getRecordFormatterCreator` method is returning default lambda with `GcsBigQueryDenormalizedRecordFormatter` so it's just testing that it returned expected value.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java (hunk 2 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:133]: I extract some part but setup is not always same, it has some differences

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java (hunk 3 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java:28]: I think you are right but should I add this change in this PR?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java (hunk 5 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java:263]: In the previous implementation this was testing `DefaultBigQueryDenormalizedRecordFormatter` but not `GcsBigQueryDenormalizedRecordFormatter` so if you would check `DefaultBigQueryDenormalizedRecordFormatterTest` there you could see this assert with expecting "2020-10-14 01:06:29.000000+00:00"

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/DefaultArrayFormatterTest.java (hunk 6 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/DefaultArrayFormatterTest.java:40]: Yeap, it's more visible

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java:48]: nice, makes sense

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java]: yeah it feels redundant to test both of those, and I'd prefer to remove this one...

I'm OK with having a single test case here just to satisfy the coverage counter, but would be nice to just not have this one. (I'm assuming that the actual NamingResolver has much more extensive testing, too)

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java [hunk 13]
```diff
@@ -109,7 +116,17 @@ void isDefaultAirbyteTmpTableSchema() {
 
   @Test
   void getAvroSchemaCreator() {
-    assertNotNull(bqdd.getAvroSchemaCreator());
+    DefaultBigQueryRecordFormatter recordFormatterMock = mock(DefaultBigQueryRecordFormatter.class);
+    AirbyteStreamNameNamespacePair namespacePairMock = mock(AirbyteStreamNameNamespacePair.class);
+    when(recordFormatterMock.getJsonSchema()).thenReturn(mock(JsonNode.class));
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:121]: nitpick: generally avoid mocking pure data objects; instead just return `mapper.createObjectNode()`

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java]: honestly, this test has kind of the same problem as the naming resolver test, where we're basically just double-verifying the behavior of some other class....

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java (hunk 3 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java:28]: probably not - can you submit an issue for this?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java (hunk 4 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java:80]: how about just updating the first and last test methods in each class? I think that would help future code authors make good choices when they add new tests, without requiring a rewrite on every test method.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java (hunk 5 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java:263]: wait, so the record formatters actually behave differently? that sounds like a bug, unless I'm misunderstanding?

(why do we need multiple record formatters anyway?)

**Review by edgao — COMMENTED** [review]
left a few inline comments - I think this is really close. Couple final questions and then we should be good to merge 🚛

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java (hunk 1 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:187]: the test is for `putStreamIntoUploaderMap` method functionality so it check the logic of this method. In this logic we are setting 
`((DefaultBigQueryDenormalizedRecordFormatter) formatter).setArrayFormatter(new LegacyArrayFormatter());`
or no. 
We can check if it use the default formatter instead but then I suppose it will be similar to integration test.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java (hunk 3 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/JsonSchemaFormatTest.java:28]: Done

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java (hunk 5 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/GcsBigQueryDenormalizedRecordFormatterTest.java:263]: There is some difference between `DefaultBigQueryDenormalizedRecordFormatter` and `GcsBigQueryDenormalizedRecordFormatter`. For example in this case GCS s use MILLISECONDS but Default formatter MICROSECONDS. So the GCS just add small extra functionality on the top of default.

What I think is the tests for GCS formatter were designed wrong since it test the Default formatter functionality. What I propose is to leave tests for Default one as it is but adapt a bit GCS tests to cover it's own functionality only. Basically it's some replacement in json schema and this MILLISECONDS thing. 

What do you think?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java (hunk 4 above)
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatterTest.java]: I made a refactoring for Default and GCS formate's tests. 
- expected schemas are used now 
- now`GcsBigQueryDenormalizedRecordFormatterTest` cover only it's own functionality

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (suhomud) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java]: I agree. Maybe we have to discuss this requirement of 90 % unit test coverage. I mean for me the only purpose of unit test is test a particular functionality like methods and it's not meaningful to cover the method which just call another method

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java (hunk 1 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:159]: ah yeah, that makes sense. I think this is fine as-is.

**Review by edgao — APPROVED** [review]
:shipit: thanks for dealing with all the back-and-forth here! LGTM :tada:
